### PR TITLE
Modify OpHit finder algorithms to save the StartTime/RiseTime

### PR DESCRIPTION
**Pull request by fjnicolas**
This draft PR (i) modifies the OpHit finder to save the StartTime and the RiseTime new attributes added in LarSoft/lardataobj#26 and (ii) adds an art tool to calculate the RiseTime for the reconstructed pulses. Changes are summarized below.

- [OpticalDetector/OpHitFinder/OpHitAlg.h](https://github.com/LArSoft/larana/compare/develop...SBNSoftware:larana:feature/fnicolas_risetime#diff-99050a3bc342a38b16ecd65b1a9a3b4b600fd52207aad08d51ba6ee2736564c4):  it now fills the StartTime and RiseTime new members. The option to save the pulse start time (overloading the PeakTime) is removed. Note this makes this PR **backwards incompatible**. To my knowledge this option is only used by `icaruscode`. Now that the OpHit data product has been modified I think it would be more convenient to update `icaruscode` than keeping the overloading option (if that makes sense to all of you). Maybe @PetrilloAtWork or @drinkingkazu can comment on that.
- [larana/OpticalDetector/OpHitFinder/PMTPulseRecoBase.h](https://github.com/LArSoft/larana/compare/develop...SBNSoftware:larana:feature/fnicolas_risetime#diff-c6aa92ac29439b1017db420129238d39338353dbdbeb4732aecdf1f49bf27401) Add `rise_time` variable to the `pulse_param` structure.
- Add a new directory `larana/OpticalDetector/OpHitFinder/RiseTimeTools/`: The RiseTime calculation is implemented through an art tool that can be called from the different pulse finder algorithms (in this PR it's implemented only for the AlgoSlidingWindow algorithm).

Presentation at coordination meeting: https://indico.fnal.gov/event/55341/contributions/246187/attachments/157023/205182/LArSoftMeeting_larana.pdf

### CODE REVIEW

File: larana/OpticalDetector/OpHitFinder/AlgoSlidingWindow.h
Comment (greenc-FNAL) [inline larana/OpticalDetector/OpHitFinder/AlgoSlidingWindow.h]: This header include should be moved to `AlgoSlidingWindow.cxx`, as that is where `make_tool()` is used. Otherwise, one is conferring an unnecessary transitive (`PUBLIC`) dependence on `art_plugin_support::toolMaker` on any includer of `AlgoSlidingWindow.h`.

File: larana/OpticalDetector/OpHitFinder/CMakeLists.txt
Comment (greenc-FNAL) [inline larana/OpticalDetector/OpHitFinder/CMakeLists.txt]: This is only a `PRIVATE` dependence if the comment on `AlgoSlidingWindow.h` is implemented.

File: larana/OpticalDetector/OpHitFinder/RiseTimeTools/CMakeLists.txt
Comment (greenc-FNAL) [inline larana/OpticalDetector/OpHitFinder/RiseTimeTools/CMakeLists.txt:11]: Although `RiseTimeCalculatorTool` does not currently confer any `INTERFACE` dependencies, it should itself still be a necessary part of any `lar::RiseTimeCalculatorTool`, viz:
```{.cmake}
  LIBRARIES CONDITIONAL larana::RiseTimeCalculatorTool
```

**Review by greenc-FNAL — CHANGES_REQUESTED** [review]
Generally a well-constructed, frugal PR.

Minor suggested changes to reduce impact on downstream code.

File: larana/OpticalDetector/OpHitFinder_module.cc
Comment (PetrilloAtWork) [inline larana/OpticalDetector/OpHitFinder_module.cc]: I vigorously oppose silently dropping this configuration option. Since who wrote a configuration including this parameter did expect the `absTime` to hold the start time, I believe we should let them know about the change.
A proposal is to emit a warning if the configuration requested `absTime` to hold the peak time (which the new configuration does anyway) and an exception if the configuration expressly requested `absTime` to be the start time.
Example (not compiled):
```cpp
    if (std::optional useStartTime = pset.get_if_present<bool>("UseStartTime")) {
      if (*useStartTime) {
        throw art::Exception{ art::errors::Configuration }
          << "OpHitFinder configuration option 'UseStartTime' is removed as recob::OpHit hosts both start and peak time.\n"
          << "See SBN DocDB 27124 for details.\n";
      }
      else {
        mf::LogWarning("OpHitFinder")
          << "OpHitFinder configuration option 'UseStartTime' is not supported any more (see SBN DocDB 27124).";
      }
    } // if "UseStartTime" in configuration
```
(needs `#include "messagefacility/MessageLogger/MessageLogger.h"` and `#include "canvas/Utilities/Exception.h"`, maybe something more)

**Review by PetrilloAtWork — CHANGES_REQUESTED** [review]
Intruding with an non-required review!
I appreciate very much the effort in adding this information to the optical hits.
I am nevertheless requesting the change to be made as loud as possible, throwing an exception when there is evidence of the intention of the user being betrayed by the new behaviour (see the comment inline).
(I have to say I would actually prefer the module to use validated configuration, but converting the whole configuration strategy of the module is beyond your duty)

**Comment by lgarren** [comment]
Thanks @PetrilloAtWork   
We discussed this PR at the coordination meeting this morning.  Since it would be a breaking change for ICARUS,  there is need for a conversation.   We need approval from ICARUS for the breaking changes.

**Comment by PetrilloAtWork** [comment]
A side comment that may be relevant: this change introduced a dependency on _art_ in one of the algorithms, which has been happily _art_-agnostic until now (that is also a constraint preventing the replacement of the hard-coded list of hit-finding and pedestal-finding algorithms by a "factory"). I don't know if there are people using the algorithm(s) outside _art_, but for sure that is how it was developed. The path of less impact is probably to preserve the old algorithm side by side with the new one.

**Comment by knoepfel** [comment]
@fjnicolas, the SciSoft team has discussed this PR at length.  Important aspects of what has already been discussed:

- ICARUS has requested that you retain the old algorithm's behavior (via a configuration parameter)
- Is it possible for a pointer to the tool to be passed into the algorithm's constructor?  In other words, the tool would be created in the module that constructs the algorithm, and the tool would be passed directly to the algorithm's constructor.  This way, the algorithm retains its independence from the art UPS product.

**Comment by fjnicolas** [comment]
Thank you all for the reviews/comments! I'm working on the changes

**Comment by fjnicolas** [comment]
Hi @knoepfel @PetrilloAtWork @lgarren, log change after the last commits:
- The legacy option to overload the PeakTime attribute (via the `fUseStartTime` fhicl parameter) is kept
- The `RiseTimeCalculator` tool is now created in the OpHit producer module and it's passed as an argument to the different pulse finder algorithm constructors
- The option to calculate the `RiseTime` is now available for all the pulse finder algorithms
- The `RiseTime` is now defined as an interval relative to the pulse `StartTime` (rather than an absolute time)

File: larana/OpticalDetector/OpHitFinder/AlgoCFD.cxx
Comment (knoepfel) [inline larana/OpticalDetector/OpHitFinder/AlgoCFD.cxx]: Recommend replacing all of this with just:

```cpp
_risetime_cal_ptr = std::move(risetimecalculator);
```

If `risetimecalculator` is `nullptr`, then `_risetime_calc_ptr` will also be `nullptr`, and then you can do... (see next comment)

File: larana/OpticalDetector/OpHitFinder/AlgoCFD.cxx
Comment (knoepfel) [inline larana/OpticalDetector/OpHitFinder/AlgoCFD.cxx]: ... (continued from previous comment)

```cpp
if (_risetime_calc_ptr)
  _pulse.t_rise = _risetime_calc_ptr->RiseTime(...);
```
In other words, the `_compute_risetime` Boolean variable is unnecessary because `_risetime_calc_ptr` will evaluate to `true` if it is non-null and `false` if is `nullptr`.

*This comment applies to the other algorithms, too.*

File: larana/OpticalDetector/OpHitFinder/PMTPulseRecoBase.h
Comment (knoepfel) [inline larana/OpticalDetector/OpHitFinder/PMTPulseRecoBase.h]: See earlier comment--this `bool` variable is unnecessary.

File: larana/OpticalDetector/OpHitFinder/RiseTimeTools/RiseTimeCalculatorBase.h
Comment (knoepfel) [inline larana/OpticalDetector/OpHitFinder/RiseTimeTools/RiseTimeCalculatorBase.h]: Recommend making this virtual function `const`-qualified (see `const` before `= 0`):

```cpp
virtual double RiseTime(const pmtana::Waveform_t& wf_pulse, const pmtana::PedestalMean_t& ped_pulse, bool _positive) const = 0;
```
You'll need to make the corresponding change to the `RiseTime(...)` overrides in the derived classes.

File: larana/OpticalDetector/OpHitFinder_module.cc
Comment (knoepfel) [inline larana/OpticalDetector/OpHitFinder_module.cc]: Yes, this is the right pattern.  However, because you have to repeat it so many times below, let's find an easier solution.  I recommend you create a function template that looks like (probably in an anonymous namespace after the `#include`s in this file):

```cpp
namespace {
  template <typename T>
  pmtana::AlgoThreshold* thresholdAlgorithm(fhicl::ParameterSet const& hit_alg_pset, 
                                            std::optional<fhicl::ParameterSet> const& rise_alg_pset)
  {
    if (rise_alg_pset) {
      return new T(hit_alg_pset, art::make_tool<RTC>(*rise_alg_pset);
    }
    return new T(hit_alg_pset, nullptr);
  }
}
```
and then you can use it this way:
```cpp
auto rise_alg_pset = pset.get_if_present<PS>("RiseTimeCalculator"); // rise_alg_pset is of type `std::optional<fhicl::ParameterSet>`
auto const hit_alg_pset = pset.get<fhicl::ParameterSet>("HitAlgoPset");
std::string threshAlgName = hit_alg_pset.get<std::string>("Name");

if (threshAlgName == "Threshold") {
  fThreshAlg = thresholdAlgorithm<pmtana::AlgoThreshold>(hit_alg_pset, rise_alg_pset);
} 
else if (threshAlgName == "SiPM") {
  fThreshAlg = thresholdAlgorithm<ptmana::AlgoSiPM>(hit_alg_pset, rise_alg_pset);
}
else if (...) ...
```

Does that make sense?

**Review by knoepfel — COMMENTED** [review]
@fjnicolas, the PR is looking good!  I have a few more comments below for your consideration.  I don't see any fundamental problems...just a few suggestions to reduce the amount of code.

Happy to chat about any of this.  And if my requests would be too time-consuming to implement, we can always consider doing those later.  But you should take these PR out of draft mode.

**Comment by fjnicolas** [comment]
Thanks @knoepfel for the last suggestions! They were implemented in the last two commits. I also removed the PR draft label.

**Review by knoepfel — COMMENTED** [review]
Looks great, @fjnicolas!  I will give L1 approval after there is L2 approval.

**Comment by lgarren** [comment]
@PetrilloAtWork Are you happy with this PR now?

**Comment by lgarren** [comment]
trigger build

**Comment by lgarren** [comment]
@fjnicolas It appears that this is an unintended breaking change for the experiments.  Please take a look at the test failures.

**Comment by fjnicolas** [comment]
@lgarren the build for SBND and ICARUS failed due to the "custom" OpHitFinder producers which make use of the old constructors for the various `PMTPulseRecoBase` algorithms ([SBNDOpHitFinder_module.cc](https://github.com/SBNSoftware/sbndcode/blob/develop/sbndcode/OpDetReco/OpHit/SBNDOpHitFinder_module.cc) and [FullOpHitFinder_module.cc](https://github.com/SBNSoftware/icaruscode/blob/develop/icaruscode/PMT/OpReco/FullOpHitFinder_module.cc)) I made the corresponding changes in `sbndcode` (see SBNSoftware/sbndcode#289, the build was successful and validation tests look good). I wasn't aware of the ICARUS OpHit producer, I'm working on the changes to make the changes backwards compatible.

There seems to be some version conflicts for the rest of the experiments (unrelated to this PR as far as I can say). I'm not sure about uBooNE fail reason, I'll have a closer look.

**Comment by lgarren** [comment]
Thanks @fjnicolas   The uboonecode problem appears to be related to this PR.    DUNE and Argoneutcode have not yet updated their packages.  I'll need to make a local test build to check their code.

**Comment by fjnicolas** [comment]
@lgarren it should build for SBND and ICARUS after the last commit

**Comment by lgarren** [comment]
trigger build

**Comment by lgarren** [comment]
@fjnicolas everything looks good except ubooncode.  See the build log failure.

`3563: /scratch/workspace/uboone_ci/label_exp/SLF7/label_exp2/swarm/uBooNE/srcs/ubana/ubana/OpticalDetectorAna/UBBasicOpticalAna_module.cc:82:31:   required from here
3564: /cvmfs/larsoft.opensciencegrid.org/products/gcc/v9_3_0/Linux64bit+3.10-2.17/include/c++/9.3.0/bits/stl_uninitialized.h:127:72: error: static assertion failed: result type must be constructible from value type of input range
3565:   127 |       static_assert(is_constructible<_ValueType2, decltype(*__first)>::value,
3566:       |                                                                        ^~~~~
`

**Comment by lgarren** [comment]
trigger build with pull requests #19, ubana@feature/fnicolas_laranahitfinder

**Comment by lgarren** [comment]
FYI: The uboone CI as built with ubana@feature/fnicolas_laranahitfinder is here:
https://dbweb8.fnal.gov:8443/LarCI/app/ns:uBooNE/view_builds/index?offset=0&builds=uboone_ci/7743&builds=

**Comment by lgarren** [comment]
approve

MicroBooNE will need ubana feature/kjk_laranahitfinder

**Pull request merged**
